File: JAVA/20_Divide&Conquer/BOJ_1891_사분면.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int d;
    static long tx, ty;
    static String str;
    static StringBuilder sb = new StringBuilder();

    public static void quadTree(long tx, long ty, long size) {
        if (size == 1) {
            System.out.println(sb);
            System.exit(0);
        }
        size /= 2;
        if (tx < size && ty >= size) {
            sb.append(1);
            quadTree(tx, ty - size, size);
        } else if (tx < size && ty < size) {
            sb.append(2);
            quadTree(tx, ty, size);
        } else if (tx >= size && ty < size) {
            sb.append(3);
            quadTree(tx - size, ty, size);
        } else if (tx >= size && ty >= size) {
            sb.append(4);
            quadTree(tx - size, ty - size, size);
        }
    }

    public static void targeted(long x, long y, long size, int idx) {
        if (idx == d) {
            tx = x;
            ty = y;
            return;
        }
        size /= 2;
        if (str.charAt(idx) == '1') {
            targeted(x, y + size, size, idx + 1);
        } else if (str.charAt(idx) == '2') {
            targeted(x, y, size, idx + 1);
        } else if (str.charAt(idx) == '3') {
            targeted(x + size, y, size, idx + 1);
        } else if (str.charAt(idx) == '4') {
            targeted(x + size, y + size, size, idx + 1);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        d = Integer.parseInt(st.nextToken());
        str = st.nextToken();
        long size = (long) Math.pow(2, d);
        targeted(0, 0, size, 0);
        st = new StringTokenizer(br.readLine(), " ");
        long y = Long.parseLong(st.nextToken());
        long x = Long.parseLong(st.nextToken());
        tx -= x;
        ty += y;
        if (tx >= 0 && tx < size && ty >= 0 && ty < size) {
            quadTree(tx, ty, size);
        } else {
            System.out.println(-1);
        }
    }
}

File: C/20_Divide&Conquer/BOJ_1891_사분면.c
#include <stdio.h>
#include <math.h>
int d;
char* str;
long long int tx, ty;
void quadTree(long long int tx, long long int ty, long long int size) {
	if (size == 1) {
		exit(0);
	}
	size /= 2;
	if (tx < size && ty >= size) {
		printf("1");
		quadTree(tx, ty - size, size);
	}
	else if (tx < size && ty < size) {
		printf("2");
		quadTree(tx, ty, size);
	}
	else if (tx >= size && ty < size) {
		printf("3");
		quadTree(tx - size, ty, size);
	}
	else if (tx >= size && ty >= size) {
		printf("4");
		quadTree(tx - size, ty - size, size);
	}
}
void targeted(long long int x, long long int y, long long int size, int idx) {
	if (idx == d) {
		tx = x;
		ty = y;
		return;
	}
	size /= 2;
	if (str[idx] == '1') {
		targeted(x, y + size, size, idx + 1);
	}
	else if (str[idx] == '2') {
		targeted(x, y, size, idx + 1);
	}
	else if (str[idx] == '3') {
		targeted(x + size, y, size, idx + 1);
	}
	else if (str[idx] == '4') {
		targeted(x + size, y + size, size, idx + 1);
	}
}
main() {
	scanf("%d", &d);
	str = (char*)malloc(sizeof(char) * d);
	scanf("%s", str);
	long long int size = pow(2, d);
	targeted(0, 0, size, 0);
	long long int x, y;
	scanf("%lld %lld", &y, &x);
	tx -= x;
	ty += y;
	if (tx >= 0 && tx < size && ty >= 0 && ty < size) {
		quadTree(tx, ty, size);
	}
	else {
		printf("-1");
	}
}